src/actions/package/providers/dnf.rs

```diff
@@ -0,0 +1,226 @@
+        vec![Step {
+            atom: Box::new(Exec {
+                command: String::from("yum"),
+                arguments: vec![
+                    String::from("install"),
+                    String::from("--assumeyes"),
+                    String::from("dnf"),
+                ],
+                privileged: true,
+                ..Default::default()
+            }),
+            initializers: vec![],
+            finalizers: vec![],
+        }]
+    }
+
+    fn has_repository(&self, _package: &PackageVariant) -> bool {
+        false
+    }
+
+    fn add_repository(&self, package: &PackageVariant) -> Vec<Step> {
+        if package.repository.is_none() {
+            return vec![];
+        }
+
+        let mut steps: Vec<Step> = vec![];
+
+        if package.key.is_some() {
+            steps.extend(vec![Step {
+                atom: Box::new(Exec {
+                    command: String::from("rpm"),
+                    arguments: vec![String::from("--import"), package.key.clone().unwrap()],
+                    privileged: true,
+                    ..Default::default()
+                }),
+                initializers: vec![],
+                finalizers: vec![],
+            }]);
+        }
+
+        steps.extend(vec![
+            Step {
+                atom: Box::new(Exec {
+                    command: String::from("dnf"),
+                    arguments: vec![
+                        String::from("config-manager"),
+                        String::from("--assumeyes"),
+                        String::from("--add-repo"),
+                        package.repository.clone().unwrap(),
+                    ],
+                    privileged: true,
+                    ..Default::default()
+                }),
+                initializers: vec![],
+                finalizers: vec![],
+            },
+            Step {
+                atom: Box::new(Exec {
+                    command: String::from("dnf"),
+                    arguments: vec![
+                        String::from("update"),
+                        String::from("--assumeyes"),
+                        String::from("--refresh"),
+                    ],
+                    privileged: true,
+                    ..Default::default()
+                }),
+                initializers: vec![],
+                finalizers: vec![],
+            },
+        ]);
+
+        steps
+    }
+
+    // TODO: What is the desired state for query(): to be package.packages() or to write custom code for each manager(like yay.rs and homebrew.rs, but not aptitude.rs)?
+    // I tried to implement query() for dnf by getting all installed packages "dnf list installed | awk '{print $1}'" and checking if they existed in package.packages,
+    // but I'm not entirely sure what the goal of query() is so for now I've left it as package.packages() and left the commented out code below
+    fn query(&self, package: &PackageVariant) -> Vec<String> {
```

Comment:
The `query` function is supposed to reduce the list of packages to the ones that need action / installed.

This is far from obvious and I will look at fixing this before 1.0

I am happy for you to return all packages, as you are, for now. Only `homebrew` gets REALLY slow when we do this, because it automatically will reinstall all packages requested.